File: 2020/2.py
import re

patt = re.compile(r'(\d+)-(\d+) (\w): (\w+)')

with open('2.txt') as f:
    lines = f.readlines()

valid = 0
for line in lines:
    if match := patt.match(line):
        if int(match.group(1)) <= match.group(4).count(match.group(3)) <= int(match.group(2)):
            valid += 1
print(valid)

valid = 0
for line in lines:
    if match := patt.match(line):
        p0, p1, c, pw = int(match.group(1)) - 1, int(match.group(2)) - 1, match.group(3), match.group(4)
        if [pw[p0], pw[p1]].count(c) == 1:
            valid += 1
print(valid)

File: 2020/2.scala
import scala.io.Source

val patt = raw"(\d+)-(\d+) (\w): (\w+)".r

val lines = Source.fromFile("2.txt").getLines().toSeq
println(lines.count {
  case patt(min, max, c, pw) => (min.toInt to max.toInt).contains(pw.count(_ == c.head))
})

println(lines.count {
  case patt(p0, p1, c, pw) => Seq(pw(p0.toInt - 1), pw(p1.toInt - 1)).count(_ == c.head) == 1
})
